### ISSUE TITLE
Import Statement uses wrong Package

### ISSUE THREAD

**Issue by nebhale** [immutables/immutables#505]
Given the following declarations:

```java
package shallow.deep;

import org.immutables.value.Value;

@Value.Immutable
abstract class AbstractCharlie {
}
```

```java
package shallow;

public abstract class Bravo<T> {
    public abstract T getEntity();
}
```

```java
package shallow.deep;

import shallow.Bravo;
import org.immutables.value.Value;

@Value.Immutable
abstract class AbstractAlpha extends Bravo<ImmutableCharlie> {
}
```

![screen shot 2016-12-13 at 13 54 59](https://cloud.githubusercontent.com/assets/60754/21160986/c3a5b0ba-c13b-11e6-9aff-13abf3bfe3a8.png)

`ImmutableAlpha` is written as

```java
package shallow.deep;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import javax.annotation.Generated;
import shallow.ImmutableCharlie;

@SuppressWarnings({"all"})
@Generated({"Immutables.generator", "AbstractAlpha"})
final class ImmutableAlpha extends AbstractAlpha {
...
```

which will not compile as `shallow.ImmutableCharlie` isn't the proper package for that type.

---

I believe that the issue is that when a generic type, derived from a generated Immutable, is encountered in the type hierarchy, it is given the package of the type that _declares_ the generic parameter, rather than the package of the type.

I've included a trivial project that exhibits this behavior: [immutables-issue.zip](https://github.com/immutables/immutables/files/650278/immutables-issue.zip)

**Comment by nebhale** [comment]
I realize that this is just a further extension of #390 but it seems to have become more problematic with an upgrade from 2.2.10 to 2.3.10.  We're working around this with FQN declarations in the generics extension declarations

```java
abstract class AbstractAlpha extends Bravo<shallow.deep.ImmutableCharlie>
```

so we've got a work around if you don't think this should be fixed.

**Comment by elucash** [comment]
Sounds very familiar, of course. I'll have to review that part again, seems that it should be easy to fix, but given that it's not ok again, the issue might something that has faded out of memory

**Comment by elucash** [comment]
released in 2.4.0, thank you!
